src/algorithms/reco/JetReconstruction.h

```diff
@@ -27,13 +30,47 @@ namespace eicrecon {
 
     public:
 
+      // algorithm initialization
       void init(std::shared_ptr<spdlog::logger> logger);
-      std::unique_ptr<edm4eic::ReconstructedParticleCollection> process(
-        const std::vector<const edm4hep::LorentzVectorE*> momenta
-      );
+
+      // run algorithm
+      template<typename T> std::unique_ptr<edm4eic::ReconstructedParticleCollection> process(const T& input_collection);
 
     private:
 
+      // generic method to add constituent to jet kinematic info
+      template <typename T> void add_to_jet_kinematics(const T& addend, edm4eic::MutableReconstructedParticle& kinematics);
+
+      // specialization for ReconstructedParticle input
+      void add_to_jet_kinematics(const edm4eic::ReconstructedParticle& addend, edm4eic::MutableReconstructedParticle& kinematics) {
+        kinematics.addToParticles(addend);
+      };
+
+      // specialization for MCParticle input
+      void add_to_jet_kinematics(const edm4hep::MCParticle& addend, edm4eic::MutableReconstructedParticle& kinematics) {
+        kinematics.addToParticles(copy_mcparticle_onto_recoparticle(addend));
+      };
+
+      // helper function to copy edm4hep::MCParticle onto edm4eic::ReconstructedParticle
+      // FIXME there is likely a far more elegant way of doing this...
+      edm4eic::ReconstructedParticle copy_mcparticle_onto_recoparticle(const edm4hep::MCParticle& mc) {
+
+        // calculate total energy
+        const float mom_squared = (std::pow(mc.getMomentum().x, 2) + std::pow(mc.getMomentum().y, 2) + std::pow(mc.getMomentum().z, 2));
+        const float energy = std::sqrt(std::pow(mc.getMass(), 2) + mom_squared);
+
+        // create reco particle
+        edm4eic::MutableReconstructedParticle mutable_reco;
```

Comment:
Does this even work? This particle is not added to any PODIO collection that is in the frame, that should result in an unusable object reference when saved to a file. Not quite sure yet what to advise here.

Reply:
TBH I'm not sure if this actually works yet (it probably doesn't) 😅 I'll be checking that soon.

This probably isn't the right way to go about it, but I couldn't think of another way to add an MCParticle to a ReconstructedParticle...